File: frontend/contratos/lib/app/data/model/contrato_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:contratos/app/infra/infra_imports.dart';
import 'package:contratos/app/data/model/model_imports.dart';
import 'package:intl/intl.dart';

class ContratoModel {
	int? id;
	int? idSolicitacaoServico;
	int? idTipoContrato;
	String? numero;
	String? nome;
	String? descricao;
	DateTime? dataCadastro;
	DateTime? dataInicioVigencia;
	DateTime? dataFimVigencia;
	String? diaFaturamento;
	double? valor;
	int? quantidadeParcelas;
	int? intervaloEntreParcelas;
	String? classificacaoContabilConta;
	String? observacao;
	List<ContratoHistoricoReajusteModel>? contratoHistoricoReajusteModelList;
	List<ContratoPrevFaturamentoModel>? contratoPrevFaturamentoModelList;
	List<ContratoHistFaturamentoModel>? contratoHistFaturamentoModelList;
	TipoContratoModel? tipoContratoModel;
	ContratoSolicitacaoServicoModel? contratoSolicitacaoServicoModel;

	ContratoModel({
		this.id,
		this.idSolicitacaoServico,
		this.idTipoContrato,
		this.numero,
		this.nome,
		this.descricao,
		this.dataCadastro,
		this.dataInicioVigencia,
		this.dataFimVigencia,
		this.diaFaturamento,
		this.valor,
		this.quantidadeParcelas,
		this.intervaloEntreParcelas,
		this.classificacaoContabilConta,
		this.observacao,
		this.contratoHistoricoReajusteModelList,
		this.contratoPrevFaturamentoModelList,
		this.contratoHistFaturamentoModelList,
		this.tipoContratoModel,
		this.contratoSolicitacaoServicoModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'numero',
		'nome',
		'descricao',
		'data_cadastro',
		'data_inicio_vigencia',
		'data_fim_vigencia',
		'dia_faturamento',
		'valor',
		'quantidade_parcelas',
		'intervalo_entre_parcelas',
		'classificacao_contabil_conta',
		'observacao',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Numero',
		'Nome',
		'Descricao',
		'Data Cadastro',
		'Data Inicio Vigencia',
		'Data Fim Vigencia',
		'Dia Faturamento',
		'Valor',
		'Quantidade Parcelas',
		'Intervalo Entre Parcelas',
		'Classificacao Contabil Conta',
		'Observacao',
	];

	ContratoModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idSolicitacaoServico = jsonData['idSolicitacaoServico'];
		idTipoContrato = jsonData['idTipoContrato'];
		numero = jsonData['numero'];
		nome = jsonData['nome'];
		descricao = jsonData['descricao'];
		dataCadastro = jsonData['dataCadastro'] != null ? DateTime.tryParse(jsonData['dataCadastro']) : null;
		dataInicioVigencia = jsonData['dataInicioVigencia'] != null ? DateTime.tryParse(jsonData['dataInicioVigencia']) : null;
		dataFimVigencia = jsonData['dataFimVigencia'] != null ? DateTime.tryParse(jsonData['dataFimVigencia']) : null;
		diaFaturamento = jsonData['diaFaturamento'];
		valor = jsonData['valor']?.toDouble();
		quantidadeParcelas = jsonData['quantidadeParcelas'];
		intervaloEntreParcelas = jsonData['intervaloEntreParcelas'];
		classificacaoContabilConta = jsonData['classificacaoContabilConta'];
		observacao = jsonData['observacao'];
		contratoHistoricoReajusteModelList = (jsonData['contratoHistoricoReajusteModelList'] as Iterable?)?.map((m) => ContratoHistoricoReajusteModel.fromJson(m)).toList() ?? [];
		contratoPrevFaturamentoModelList = (jsonData['contratoPrevFaturamentoModelList'] as Iterable?)?.map((m) => ContratoPrevFaturamentoModel.fromJson(m)).toList() ?? [];
		contratoHistFaturamentoModelList = (jsonData['contratoHistFaturamentoModelList'] as Iterable?)?.map((m) => ContratoHistFaturamentoModel.fromJson(m)).toList() ?? [];
		tipoContratoModel = jsonData['tipoContratoModel'] == null ? TipoContratoModel() : TipoContratoModel.fromJson(jsonData['tipoContratoModel']);
		contratoSolicitacaoServicoModel = jsonData['contratoSolicitacaoServicoModel'] == null ? ContratoSolicitacaoServicoModel() : ContratoSolicitacaoServicoModel.fromJson(jsonData['contratoSolicitacaoServicoModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idSolicitacaoServico'] = idSolicitacaoServico != 0 ? idSolicitacaoServico : null;
		jsonData['idTipoContrato'] = idTipoContrato != 0 ? idTipoContrato : null;
		jsonData['numero'] = numero;
		jsonData['nome'] = nome;
		jsonData['descricao'] = descricao;
		jsonData['dataCadastro'] = dataCadastro != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataCadastro!) : null;
		jsonData['dataInicioVigencia'] = dataInicioVigencia != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataInicioVigencia!) : null;
		jsonData['dataFimVigencia'] = dataFimVigencia != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataFimVigencia!) : null;
		jsonData['diaFaturamento'] = diaFaturamento;
		jsonData['valor'] = valor;
		jsonData['quantidadeParcelas'] = quantidadeParcelas;
		jsonData['intervaloEntreParcelas'] = intervaloEntreParcelas;
		jsonData['classificacaoContabilConta'] = classificacaoContabilConta;
		jsonData['observacao'] = observacao;
		
		var contratoHistoricoReajusteModelLocalList = []; 
		for (ContratoHistoricoReajusteModel object in contratoHistoricoReajusteModelList ?? []) { 
			contratoHistoricoReajusteModelLocalList.add(object.toJson); 
		}
		jsonData['contratoHistoricoReajusteModelList'] = contratoHistoricoReajusteModelLocalList;
		
		var contratoPrevFaturamentoModelLocalList = []; 
		for (ContratoPrevFaturamentoModel object in contratoPrevFaturamentoModelList ?? []) { 
			contratoPrevFaturamentoModelLocalList.add(object.toJson); 
		}
		jsonData['contratoPrevFaturamentoModelList'] = contratoPrevFaturamentoModelLocalList;
		
		var contratoHistFaturamentoModelLocalList = []; 
		for (ContratoHistFaturamentoModel object in contratoHistFaturamentoModelList ?? []) { 
			contratoHistFaturamentoModelLocalList.add(object.toJson); 
		}
		jsonData['contratoHistFaturamentoModelList'] = contratoHistFaturamentoModelLocalList;
		jsonData['tipoContratoModel'] = tipoContratoModel?.toJson;
		jsonData['contratoSolicitacaoServicoModel'] = contratoSolicitacaoServicoModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idSolicitacaoServico = plutoRow.cells['idSolicitacaoServico']?.value;
		idTipoContrato = plutoRow.cells['idTipoContrato']?.value;
		numero = plutoRow.cells['numero']?.value;
		nome = plutoRow.cells['nome']?.value;
		descricao = plutoRow.cells['descricao']?.value;
		dataCadastro = Util.stringToDate(plutoRow.cells['dataCadastro']?.value);
		dataInicioVigencia = Util.stringToDate(plutoRow.cells['dataInicioVigencia']?.value);
		dataFimVigencia = Util.stringToDate(plutoRow.cells['dataFimVigencia']?.value);
		diaFaturamento = plutoRow.cells['diaFaturamento']?.value;
		valor = plutoRow.cells['valor']?.value?.toDouble();
		quantidadeParcelas = plutoRow.cells['quantidadeParcelas']?.value;
		intervaloEntreParcelas = plutoRow.cells['intervaloEntreParcelas']?.value;
		classificacaoContabilConta = plutoRow.cells['classificacaoContabilConta']?.value;
		observacao = plutoRow.cells['observacao']?.value;
		contratoHistoricoReajusteModelList = [];
		contratoPrevFaturamentoModelList = [];
		contratoHistFaturamentoModelList = [];
		tipoContratoModel = TipoContratoModel();
		tipoContratoModel?.nome = plutoRow.cells['tipoContratoModel']?.value;
		contratoSolicitacaoServicoModel = ContratoSolicitacaoServicoModel();
		contratoSolicitacaoServicoModel?.descricao = plutoRow.cells['contratoSolicitacaoServicoModel']?.value;
	}	

	ContratoModel clone() {
		return ContratoModel(
			id: id,
			idSolicitacaoServico: idSolicitacaoServico,
			idTipoContrato: idTipoContrato,
			numero: numero,
			nome: nome,
			descricao: descricao,
			dataCadastro: dataCadastro,
			dataInicioVigencia: dataInicioVigencia,
			dataFimVigencia: dataFimVigencia,
			diaFaturamento: diaFaturamento,
			valor: valor,
			quantidadeParcelas: quantidadeParcelas,
			intervaloEntreParcelas: intervaloEntreParcelas,
			classificacaoContabilConta: classificacaoContabilConta,
			observacao: observacao,
			contratoHistoricoReajusteModelList: contratoHistoricoReajusteModelListClone(contratoHistoricoReajusteModelList!),
			contratoPrevFaturamentoModelList: contratoPrevFaturamentoModelListClone(contratoPrevFaturamentoModelList!),
			contratoHistFaturamentoModelList: contratoHistFaturamentoModelListClone(contratoHistFaturamentoModelList!),
		);			
	}

	contratoHistoricoReajusteModelListClone(List<ContratoHistoricoReajusteModel> contratoHistoricoReajusteModelList) { 
		List<ContratoHistoricoReajusteModel> resultList = [];
		for (var contratoHistoricoReajusteModel in contratoHistoricoReajusteModelList) {
			resultList.add(
				ContratoHistoricoReajusteModel(
					id: contratoHistoricoReajusteModel.id,
					idContrato: contratoHistoricoReajusteModel.idContrato,
					indice: contratoHistoricoReajusteModel.indice,
					valorAnterior: contratoHistoricoReajusteModel.valorAnterior,
					valorAtual: contratoHistoricoReajusteModel.valorAtual,
					dataReajuste: contratoHistoricoReajusteModel.dataReajuste,
					observacao: contratoHistoricoReajusteModel.observacao,
				)
			);
		}
		return resultList;
	}

	contratoPrevFaturamentoModelListClone(List<ContratoPrevFaturamentoModel> contratoPrevFaturamentoModelList) { 
		List<ContratoPrevFaturamentoModel> resultList = [];
		for (var contratoPrevFaturamentoModel in contratoPrevFaturamentoModelList) {
			resultList.add(
				ContratoPrevFaturamentoModel(
					id: contratoPrevFaturamentoModel.id,
					idContrato: contratoPrevFaturamentoModel.idContrato,
					dataPrevista: contratoPrevFaturamentoModel.dataPrevista,
					valor: contratoPrevFaturamentoModel.valor,
				)
			);
		}
		return resultList;
	}

	contratoHistFaturamentoModelListClone(List<ContratoHistFaturamentoModel> contratoHistFaturamentoModelList) { 
		List<ContratoHistFaturamentoModel> resultList = [];
		for (var contratoHistFaturamentoModel in contratoHistFaturamentoModelList) {
			resultList.add(
				ContratoHistFaturamentoModel(
					id: contratoHistFaturamentoModel.id,
					idContrato: contratoHistFaturamentoModel.idContrato,
					dataFatura: contratoHistFaturamentoModel.dataFatura,
					valor: contratoHistFaturamentoModel.valor,
				)
			);
		}
		return resultList;
	}

	
}

File: backend/contratos/src/model/contrato_model.py
from src import db
from src.model.contrato_historico_reajuste_model import ContratoHistoricoReajusteModel
from src.model.contrato_prev_faturamento_model import ContratoPrevFaturamentoModel
from src.model.contrato_hist_faturamento_model import ContratoHistFaturamentoModel
from src.model.tipo_contrato_model import TipoContratoModel
from src.model.contrato_solicitacao_servico_model import ContratoSolicitacaoServicoModel


class ContratoModel(db.Model):
    __tablename__ = 'contrato'

    id = db.Column(db.Integer, primary_key=True)
    numero = db.Column(db.String(50))
    nome = db.Column(db.String(100))
    descricao = db.Column(db.Text)
    data_cadastro = db.Column(db.DateTime)
    data_inicio_vigencia = db.Column(db.DateTime)
    data_fim_vigencia = db.Column(db.DateTime)
    dia_faturamento = db.Column(db.String(2))
    valor = db.Column(db.Float)
    quantidade_parcelas = db.Column(db.Integer)
    intervalo_entre_parcelas = db.Column(db.Integer)
    classificacao_contabil_conta = db.Column(db.String(30))
    observacao = db.Column(db.Text)
    id_tipo_contrato = db.Column(db.Integer, db.ForeignKey('tipo_contrato.id'))
    id_solicitacao_servico = db.Column(db.Integer, db.ForeignKey('contrato_solicitacao_servico.id'))

    contrato_historico_reajuste_model_list = db.relationship('ContratoHistoricoReajusteModel', lazy='dynamic')
    contrato_prev_faturamento_model_list = db.relationship('ContratoPrevFaturamentoModel', lazy='dynamic')
    contrato_hist_faturamento_model_list = db.relationship('ContratoHistFaturamentoModel', lazy='dynamic')
    tipo_contrato_model = db.relationship('TipoContratoModel', foreign_keys=[id_tipo_contrato])
    contrato_solicitacao_servico_model = db.relationship('ContratoSolicitacaoServicoModel', foreign_keys=[id_solicitacao_servico])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_solicitacao_servico = data.get('idSolicitacaoServico')
        self.id_tipo_contrato = data.get('idTipoContrato')
        self.numero = data.get('numero')
        self.nome = data.get('nome')
        self.descricao = data.get('descricao')
        self.data_cadastro = data.get('dataCadastro')
        self.data_inicio_vigencia = data.get('dataInicioVigencia')
        self.data_fim_vigencia = data.get('dataFimVigencia')
        self.dia_faturamento = data.get('diaFaturamento')
        self.valor = data.get('valor')
        self.quantidade_parcelas = data.get('quantidadeParcelas')
        self.intervalo_entre_parcelas = data.get('intervaloEntreParcelas')
        self.classificacao_contabil_conta = data.get('classificacaoContabilConta')
        self.observacao = data.get('observacao')

    def serialize(self):
        return {
            'id': self.id,
            'idSolicitacaoServico': self.id_solicitacao_servico,
            'idTipoContrato': self.id_tipo_contrato,
            'numero': self.numero,
            'nome': self.nome,
            'descricao': self.descricao,
            'dataCadastro': self.data_cadastro.isoformat(),
            'dataInicioVigencia': self.data_inicio_vigencia.isoformat(),
            'dataFimVigencia': self.data_fim_vigencia.isoformat(),
            'diaFaturamento': self.dia_faturamento,
            'valor': self.valor,
            'quantidadeParcelas': self.quantidade_parcelas,
            'intervaloEntreParcelas': self.intervalo_entre_parcelas,
            'classificacaoContabilConta': self.classificacao_contabil_conta,
            'observacao': self.observacao,
            'contratoHistoricoReajusteModelList': [contrato_historico_reajuste_model.serialize() for contrato_historico_reajuste_model in self.contrato_historico_reajuste_model_list],
            'contratoPrevFaturamentoModelList': [contrato_prev_faturamento_model.serialize() for contrato_prev_faturamento_model in self.contrato_prev_faturamento_model_list],
            'contratoHistFaturamentoModelList': [contrato_hist_faturamento_model.serialize() for contrato_hist_faturamento_model in self.contrato_hist_faturamento_model_list],
            'tipoContratoModel': self.tipo_contrato_model.serialize() if self.tipo_contrato_model else None,
            'contratoSolicitacaoServicoModel': self.contrato_solicitacao_servico_model.serialize() if self.contrato_solicitacao_servico_model else None,
        }